src/main/java/org/opensearch/knn/index/codec/util/KNNCodecUtil.java

```diff
@@ -38,32 +38,52 @@ public static final class Pair {
         @Setter
         private int dimension;
         public SerializationMode serializationMode;
+        public boolean finished;
     }
 
     /**
      * Extract docIds and vectors from binary doc values.
-     *
-     * @param values Binary doc values
-     * @param vectorTransfer Utility to make transfer
-     * @return KNNCodecUtil.Pair representing doc ids and corresponding vectors
-     * @throws IOException thrown when unable to get binary of vectors
-     */
-    public static KNNCodecUtil.Pair getPair(final BinaryDocValues values, final VectorTransfer vectorTransfer) throws IOException {
+    *
+    * @param values Binary doc values
+    * @param vectorTransfer Utility to make transfer
+    * @return KNNCodecUtil.Pair representing doc ids and corresponding vectors
+    * @throws IOException thrown when unable to get binary of vectors
+    */
+    public static KNNCodecUtil.VectorBatch getVectorBatch(
+        final BinaryDocValues values,
+        final VectorTransfer vectorTransfer,
+        boolean iterative
+    ) throws IOException {
         List<Integer> docIdList = new ArrayList<>();
         SerializationMode serializationMode = SerializationMode.COLLECTION_OF_FLOATS;
-        vectorTransfer.init(getTotalLiveDocsCount(values));
-        for (int doc = values.nextDoc(); doc != DocIdSetIterator.NO_MORE_DOCS; doc = values.nextDoc()) {
+        if (iterative) {
+            // Initializing with a value of zero means to only allocate as much memory on JNI as
+            // we have inserted for vectors in java side
+            vectorTransfer.init(0);
+        } else {
+            vectorTransfer.init(getTotalLiveDocsCount(values));
+        }
+        int doc = values.nextDoc();
+        for (; doc != DocIdSetIterator.NO_MORE_DOCS; doc = values.nextDoc()) {
             BytesRef bytesref = values.binaryValue();
             serializationMode = vectorTransfer.getSerializationMode(bytesref);
             vectorTransfer.transfer(bytesref);
             docIdList.add(doc);
+            // Semi-hacky way to check if the streaming limit has been reached
+            if (iterative && vectorTransfer.getVectorAddress() != 0) {
```

Comment:
Are we transferring one vector at a time if its iterative? Can we do in batches? Also, instead of the iterative arg, can we just have how many vectors should be fetched? That way, this function does not need to be concerned with what the vectors are being used for.

Reply:
If the iterative flag is set to true, this means to return after we have reached vectorStreamingLimit worth of vectors. Maybe it would be best to rename the flag to reduce confusion? We could also theoretically change this to how big the batch size is, but I'm not sure that would be better. This flag accounts for the two cases of:
1. Push all of the vectors into memory
2. Push one batch with specified size into memory.